### PR TITLE
EVG-3230 Add retries and timeout to create host

### PR DESCRIPTION
**Pull request by bsamek**

### CODE REVIEW

File: units/provisioning_create_host.go [hunk 1]
```diff
@@ -73,66 +90,97 @@ func (j *createHostJob) Run(ctx context.Context) {
...
-	// Provisionally set h.StartTime to now. Cloud providers may override
+	// Provisionally set j.host.StartTime to now. Cloud providers may override
 	// this value with the time the host was created.
-	h.StartTime = time.Now()
+	j.host.StartTime = j.start
 
-	_, err = h.Upsert()
-	if err != nil {
-		return errors.Wrapf(err, "error updating host %v", h.Id)
+	if err := j.host.Insert(); err != nil {
```
Comment (john-m-liu) [inline units/provisioning_create_host.go]: Won't this hit duplicate key errors?

File: model/host/host.go [hunk 2]
```diff
@@ -144,6 +144,9 @@ type SpawnOptions struct {
 
 	// Retries is the number of times Evergreen should try to spawn this host.
 	Retries int `bson:"retries,omitempty" json:"retries,omitempty"`
+
+	// SpawnedByTask indicates that this host has been spawned by a task.
+	SpawnedByTask bool `bson:"spawned_by_task,omitempty" json:"spawned_by_task,omitempty"`
```
[the remark this answers is not in the document]
Comment (bsamek) [reply model/host/host.go]: No, TaskID is only set if it's a task-scoped host. We could check both TaskID and BuildID but I liked this a little bit more.

File: units/provisioning_create_host.go [hunk 3]
```diff
@@ -47,20 +50,34 @@ func makeCreateHostJob() *createHostJob {
 	return j
 }
 
-func NewHostCreateJob(env evergreen.Environment, h host.Host, id string) amboy.Job {
+func NewHostCreateJob(ctx context.Context, env evergreen.Environment, h host.Host, id string, CurrentAttempt int, MaxAttempts int) amboy.Job {
```
[the remark this answers is not in the document]
Comment (bsamek) [reply units/provisioning_create_host.go]: Done

File: units/provisioning_create_host.go (hunk 1 above)
Comment (bsamek) [reply units/provisioning_create_host.go]: No, if spawning a host fails, this function returns on 139. It only inserts the host once, i.e., if it succeeds.

File: units/provisioning_create_host.go [hunk 4]
```diff
@@ -47,20 +50,34 @@ func makeCreateHostJob() *createHostJob {
 	return j
 }
 
-func NewHostCreateJob(env evergreen.Environment, h host.Host, id string) amboy.Job {
+func NewHostCreateJob(env evergreen.Environment, h host.Host, id string, CurrentAttempt int, MaxAttempts int) amboy.Job {
```
Comment (tychoish) [inline units/provisioning_create_host.go]: I think it's really easy (and not clear from the call site) to flip successive integer elements. Does it make sense to define a (util) "RetryCounter" type.

File: units/provisioning_create_host.go (hunk 1 above)
Comment (tychoish) [reply units/provisioning_create_host.go]: there's a db.IsDuplicateKey(err error) function that you can use to return nil in this case, which is probably what you want (for belt and suspenders, we should also change that function to use errors.Cause()

**Pull request merged**